daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionSpec.scala

```diff
@@ -675,6 +675,363 @@ class TransactionSpec
         Map(key("key0") -> Some(cid0), key("key1") -> None, key("key2") -> Some(cid3))
     }
   }
+
+  "consumedBy and rolledbackBy" - {
+    "non-consuming transaction with no rollbacks" - {
+      "no nodes" in {
+        val builder = TransactionBuilder()
+        val transaction = builder.build()
+
+        transaction.consumedBy shouldBe Map.empty
+        transaction.rolledbackBy shouldBe Map.empty
+      }
+
+      "one node" - {
+        "with local contracts" in {
+          val builder = TransactionBuilder()
+          val parties = Seq("Alice")
+          val (_, createNode0) = create(builder, parties, Some("key0"))
+
+          builder.add(createNode0)
+          val transaction = builder.build()
+
+          transaction.consumedBy shouldBe Map.empty
+          transaction.rolledbackBy shouldBe Map.empty
+        }
+
+        "with global contracts" in {
+          val builder = TransactionBuilder()
+          val parties = Seq("Alice")
+          val (_, createNode0) = create(builder, parties, Some("key0"))
+          val fetchNode0 = builder.fetch(createNode0, true)
+
+          builder.add(fetchNode0)
+          val transaction = builder.build()
+
+          transaction.consumedBy shouldBe Map.empty
+          transaction.rolledbackBy shouldBe Map.empty
+        }
+      }
+
+      "multiple nodes" - {
+        "only create nodes" in {
+          val builder = TransactionBuilder()
+          val parties = Seq("Alice")
+          val (_, createNode0) = create(builder, parties, Some("key0"))
+          val (_, createNode1) = create(builder, parties, Some("key1"))
+
+          builder.add(createNode0)
+          builder.add(createNode1)
+          val transaction = builder.build()
+
+          transaction.consumedBy shouldBe Map.empty
+          transaction.rolledbackBy shouldBe Map.empty
+        }
+
+        "create and non-consuming exercise nodes" - {
+          "with local contracts" in {
+            val builder = TransactionBuilder()
+            val parties = Seq("Alice")
+            val (_, createNode0) = create(builder, parties, Some("key0"))
+
+            builder.add(createNode0)
+            builder.add(exercise(builder, createNode0, parties, false))
+            val transaction = builder.build()
+
+            transaction.consumedBy shouldBe Map.empty
+            transaction.rolledbackBy shouldBe Map.empty
+          }
+
+          "with global contracts" in {
+            val builder = TransactionBuilder()
+            val parties = Seq("Alice")
+            val (_, createNode0) = create(builder, parties, Some("key0"))
+
+            builder.add(exercise(builder, createNode0, parties, false))
```

Comment:
where is the test that a consuming exercise of a global contract _does_ get include? This only checks that a non-consuming exercise does not get included.

Reply:
These tests (lines 680-759) are only concerned with non-consuming transactions (no rollbacks).

The tests (lines 760-824) are concerned with consuming transactions (no rollbacks).

So, an example test for a consuming exercise of a global contract should be at lines 776-787 - i.e.
```
          val builder = TransactionBuilder()
          val parties = Seq("Alice")
          val (cid0, createNode0) = create(builder, parties, Some("key0"))

          val exerciseId0 = builder.add(exercise(builder, createNode0, parties, true))
          val transaction = builder.build()

          transaction.consumedBy shouldBe
            Map(cid0 -> exerciseId0)
          transaction.rolledbackBy shouldBe Map.empty
```

Reply:
makes sense, thanks!